aws/resource_aws_api_gateway_authorizer.go

```diff
@@ -21,7 +21,7 @@ func resourceAwsApiGatewayAuthorizer() *schema.Resource {
 		Schema: map[string]*schema.Schema{
 			"authorizer_uri": &schema.Schema{
 				Type:     schema.TypeString,
-				Required: true,
+				Optional: true,
```

Comment:
Sorry I'm unfamiliar with this resource, but is this required for `TOKEN`/`REQUEST`? Just curious if that was an oversight before, if we should add documentation, or `CustomizeDiff` for continued plan-time validation

Reply:
@bflad Indeed `CustomizeDiff` should be used here. But it doesn't seem to get the argument. Did I miss something? 
https://github.com/loivis/terraform-provider-aws/blob/9726a5187a1fb34a8edbc660e0d3968d8e46a3f0/aws/resource_aws_api_gateway_authorizer.go#L277

Reply:
I think you need to perform two checks, one for existence and one for string length:

```go
v, ok := diff.GetOk("authorizer_uri")
if ok && v.(string) == "" {
  return errors.New("authorizer_uri must be non-empty when type is ...")
} else {
  return errors.New("authorizer_uri is required when type is ...")
}
```

You'll need to do the same with `len() == 0` for provider_arns

Reply:
I did try that. The problem is that `diff.GetOk("authorizer_uri")` didn't give me `true` for `ok`.

Reply:
I added a few lines to debug
```
	args := []string{"authorizer_uri", "name", "rest_api_id", "identity_source", "type", "identity_validation_expression", "authorizer_credentials"}
	for _, arg := range args {
		val, ok := diff.GetOk(arg)
		log.Printf("[DEBUG] %s: #%s#, #%v#", arg, val.(string), ok)
	}
```
It gives me logs:
```
2018/03/05 21:50:54 [DEBUG] authorizer_uri: ##, #false#
2018/03/05 21:50:54 [DEBUG] name: #tf-acctest-igw-authorizer-yddwa40#, #true#
2018/03/05 21:50:54 [DEBUG] rest_api_id: ##, #false#
2018/03/05 21:50:54 [DEBUG] identity_source: #method.request.header.Authorization#, #true#
2018/03/05 21:50:54 [DEBUG] type: #TOKEN#, #true#
2018/03/05 21:50:54 [DEBUG] identity_validation_expression: ##, #false#
2018/03/05 21:50:54 [DEBUG] authorizer_credentials: ##, #false#
```
`name`, `identity_source`, `type` are arguments with default value. It seems like `diff.GetOk()` doesn't return value for arguments without default value. This is what makes me confused.

Reply:
Ah, this is happening because the values are computed:

https://github.com/hashicorp/terraform/blob/ea50f455edca3acaa0a6ab372b861e6cd2d386aa/helper/schema/resource_diff.go#L339

Unfortunately, I'm not sure of a workaround at the moment. For now, we can keep that logic outside `CustomizeDiff`. 🙁